ql/src/java/org/apache/hadoop/hive/ql/optimizer/ConvertJoinMapJoin.java

```diff
@@ -755,6 +757,20 @@ private boolean checkConvertJoinSMBJoin(JoinOperator joinOp, OptimizeTezProcCont
         LOG.debug("External table {} found in join and also could not provide statistics - disabling SMB join.", sb);
         return false;
       }
+      // GBY operators buffers one record. These are processed when ReduceRecordSources flushes the operator tree
+      // when end of record stream reached. If the tree has more than two GBY operators CommonMergeJoinOperator can
+      // not process all buffered records.
+      // HIVE-27788
+      if (parentOp.getParentOperators() != null) {
+        for (Operator<?> grandParent : parentOp.getParentOperators()) {
```

Comment:
Before the SMB conversion the direct parents of the merge join operator are RS operators. Passing them to `hasMoreOperatorsThan` would immediately stop traversal because RS operator is a stop operator in this search.

Reply:
Thanks for the clarification. Maybe leave a comment or assertion that parentOp is always an RS.